### ISSUE TITLE
Semantic tokens: When is SemanticTokensRequestOptions given?

### ISSUE THREAD

**Issue by Tyriar** [microsoft/vscode#86167]
Testing #85976

In my testing on a TS file, the options arg in provideSemanticTokens was always `{}`, regardless of whether a new line was added/deleted or a change was made on a single line. When are options given?

**Comment by alexandrudima** [comment]
#### `SemanticTokensRequestOptions.ranges`
This option is currently not exercised in the implementation. Here is my thinking for when we would use it.

When a file is opened, we would do two requests:
- the first request `A` for semantic tokens would be with `ranges` set to the current viewport. The language server would first see this request and can answer with a small subset of tokens (only the ones in the requested ranges). We don't do this request today.
- immediately after the first request, we would make the request `B` without `ranges`, which means for the entire file. This is the request we do today.
- if the user scrolls in the editor, and `B` is still running (not yet resolved), we would make more requests for the new viewport and cancel older ones if the scroll position is changed and there is no more overlap... But we wouldn't cancel `B`...
- once `B` has resolved, and the editor has all the tokens in the entire file, then it can scroll to any position and render semantic tokens immediately.
- so once `B` has resolved, the problem changes from fetching semantic tokens to keeping the tokens the editor has up-to-date while editing the file.
- given our delta-encoding of tokens, I think that in most cases (except for worst-case scenarios like typing `/*` at the beginning of the file) a small edit will result in a small delta.
- so basically once `B` has resolved, our strategy would be to only make requests for the entire file and expect deltas...

#### `SemanticTokensRequestOptions.previousResultId`

This is exercised today and is the `resultId` in `SemanticTokens.resultId` or `SemanticTokensEdits.resultId`.

The implementation for TS that lives on the branch I pointed to is a naive implementation. It is written to always go to the `tsserver` and fetch all the tokens. That is because this delta protocol does not yet exist in `tsserver`. So the TS implementation that lives on the branch does not really exercise the `resultId`, since it never returns deltas. So `SemanticTokens.resultId` returned by TS is `undefined`, and that is why `SemanticTokensRequestOptions.previousResultId` is `undefined`.
